numpy/_core/src/multiarray/common.c

```diff
@@ -75,6 +75,12 @@ PyArray_DTypeFromObjectStringDiscovery(
         if (itemsize < 0) {
             return NULL;
         }
+        if (itemsize > NPY_MAX_INT) {
+            /* We can allow this, but should audit code paths before we do. */
+            PyErr_SetString(PyExc_TypeError,
+                    "string too large to store inside array.");
```

Comment:
Can do it, although I am a bit unenthusiastic about bothering because the alternative error is an overflow which isn't much more expressive.  (Or most likely a memory error.)

Reply:
I'm putting myself in the shoes of someone who happens to hit this error in the wild. It might not be immediately obvious especially for the unicode case that the limit is related to INT_MAX.